File: abctab2ps.py
# Copyright 2022 Curtis Penner

import os.path
import sys
import signal

import cmdline
import common
import constants
import field
import music
import parse
import tab
from log import log


args = cmdline.options()


def signal_handler():
    """ signal handler for premature termination """
    log.critical('could not install signal handler for SIGTERM and SIGINT')
    exit(130)


def is_comment(comment: str) -> bool:
    return len(comment) and comment[0] == '%'


def is_pseudo_comment(line: str) -> bool:
    return line.startswith("%%")


def is_cmdline(cmd: str) -> bool:
    """ command line """
    return len(cmd) > 2 and cmd.startswith('%!')


def process_file(fin):    # , fp_out, xref_str, pat, sel_all, search_field):
    # int type;
    common.within_tune = False
    common.within_block = False
    common.do_this_tune = False

    info = field.Field()
    # This is where there is the statements of verbose.
    # Just need to understand what the verbose number means to
    #  debug, warning, info, error, critical

    with open(fin) as f:
        lines = f.readlines()

    with open(args.output, 'w') as fp:
        if is_cmdline(lines[0].strip()):
            cmdline.process_cmdline(lines[0])
            del lines[0]

        for line in lines:
            line = line.strip()
            if not line:   # Todo: if within_block end it, otherwise do nothing
                # if common.within_tune
                continue
            if is_pseudo_comment(line):
                music.process_ps_comment(fp, line)
                continue
            if is_comment(line):
                continue

            line = parse.decomment_line(line)

            # reset_info(default_info)
            if field.is_field(line):
                info(line)

            if info.voice.parse_vocals(line):
                return

            # now parse a real line of music
            if not common.voices:
                common.voices.append(info.voice.switch_voice(constants.DEFVOICE))

            # music or tablature?
            if tab.is_tab_line(line):
                tab.parse_tab_line(line)
            else:
                parse.parse_music_line(line)

            # log.debug(f"  parsed music symbols {n_sym_0} to"
            #           f" {len(common.voices[common.ivc].syms)-1} for voice {common.ivc}")
            info.process_line(fp, line)    # field.XRef, line, xref_str, pat, sel_all)

        # if not common.epsf:
        #     buffer.buffer_eob(fp_out)
        #     buffer.write_buffer(fp_out)


def main():
    # cleanup on premature termination
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    log.debug(f"do_output: {common.do_output}")
    if common.do_output:
        log.info(f"This is abctab2ps, version {constants.VERSION}.{constants.REVISION}")

    # consolidate the filenames
    if args.filename:
        args.filenames.append(args.filename)
        log.debug(args.filenames)
    # set the page format
    common.cfmt.set_page_format()

    # common.search_field0 = args.select_field0   # default for interactive mode
    # if args.epsf:
    #     for filename in args.filename.split():
    #         name, extension = os.path.splitext(filename)

    if not args.filenames:
        # no input file specified: open stdin
        log.error("++++ No input files, read from stdin")
        sys.exit(2)

    # if common.do_output and args.make_index:
    #     log.info(f"make_index: {args.make_index}")
    #     ind = index.Index()
    #     index_file = 'index.ps'
    #     ind.open_index_file(index_file)

    # loop over files in list
    # pat, xref_str = parse.rehash_selectors(args.filenames)
    # fout = None

    # process list of input files and skip.ps and.eps files
    for filename in args.filenames:
        name, ext = os.path.splitext(filename)
        if ext == ".ps" or ext == ".eps":
            continue
        elif not ext and not os.path.exists(name):
            filename = name + '.abc'
            if not os.path.exists(filename):
                log.error(f'{filename} not found')
                continue

        log.info(f"{filename}:")
        if args.output:
            common.output = args.output
        else:
            common.output = name + '.ps'

        if not common.do_output:
            log.info(f"{filename=}: not no output")
        # parse.do_index(fin, xref_str, pat, common.select_all, args.search_field0)

        # this is the start of the process
        process_file(filename)  # fp, xref_str, pat, common.select_all, args.search_field0)

    # The rest just closes everything up
    # if not common.do_output:
    #     print(f"Selected {common.tune_num1} title {common.tune_num1} of {common.tune_num2}")

    # if common.do_output and common.make_index:
    #     index.close_index_file()
    # field.Field().close_output_file(fout)


if __name__ == '__main__':
    main()

File: c_src/abctab2ps.cpp
/*
 *  abctab2ps: a program to typeset tunes in abc format using PostScript
 *  Copyright (C) 1996 - 1998  Michael Methfessel
 *  Copyright (C) 1999 - 2004  Christoph Dalitz
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *  The author can be contacted as follows:
 *  
 *  Christoph Dalitz
 *  <christoph (dot) dalitz (at) hs-niederrhein (dot) de>
 *
 */

#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <iostream>

#include "abctab2ps.h"
#include "subs.h"
#include "format.h"
#include "util.h"
#include "parse.h"
#include "music.h"
// #include "oscompat.h"


/* ----- definitions of global variables ----- */

int db=DEBUG_LV;                  /* debug control */
int maxSyms,maxVc;                /* for malloc */
int allocSyms,allocVc;
struct ISTRUCT info,default_info; /* information fields */
char lvoiceid[STRLINFO];          /* string from last V: line */
int  nvoice,mvoice;               /* number of voices defd, nonempty */
int  ivc;                         /* current voice */
int  ivc0;                        /* top nonempty voice */
int ixpfree;                      /* first free element in xp array */
struct METERSTR default_meter;    /* data to specify the meter */
struct KEYSTR default_key;        /* data to specify the key */
                          /* things to alloc: */
struct SYMBOL  *sym;              /* symbol list */
struct SYMBOL  **symv;            /* symbols for voices */
struct XPOS    *xp;               /* shared horizontal positions */
struct VCESPEC *voice;            /* characteristics of a voice */
struct SYMBOL  **sym_st;          /* symbols a staff start */
int            *nsym_st; 

int halftones;                    /* number of halftones to transpose by */

float f0p,f5p,f1p,f0x,f5x,f1x;            /*   mapping fct */
float lnnp,bnnp,fnnp,lnnx,bnnx,fnnx;      /*   note-note spacing */
float lbnp,bbnp,rbnp,lbnx,bbnx,rbnx;      /*   bar-note spacing */
float lnbp,bnbp,rnbp,lnbx,bnbx,rnbx;      /*   note-bar spacing */


char wpool[NWPOOL];            /* pool for vocal strings */
int nwpool,nwline;             /* globals to handle wpool */

struct SYMBOL zsym;            /* symbol containing zeros */

struct FORMAT sfmt;                    /* format after initialization */
struct FORMAT dfmt;                    /* format at start of tune */
struct FORMAT cfmt;                    /* current format for output */

char fontnames[50][STRLFMT];           /* list of needed fonts */
int  nfontnames;

//char txt[MAXNTEXT][MAXWLEN];           /* for output of text */
//int  ntxt; 
StringVector words_of_text;            /* for output of text */


char vcselstr[STRLFILE];          /* string for voice selection */
char mbf[501];                 /* mini-buffer for one line */
char buf[BUFFSZ];              /* output buffer.. should hold one tune */
int nbuf;                      /* number of bytes buffered */
float bposy;                   /* current position in buffered data */
int   ln_num;                  /* number of lines in buffer */
float ln_pos[BUFFLN];          /* vertical positions of buffered lines */
int   ln_buf[BUFFLN];          /* buffer location of buffered lines */
int   use_buffer;              /* 1 if lines are being accumulated */

char text [NTEXT][STRLINFO];   /* pool for history, words, etc. lines */
int text_type[NTEXT];          /* type of each text line */
int ntext;                     /* number of text lines */
char page_init[201];           /* initialization string after page break */
bool do_output;                // do index (false) or output (true)
char escseq[81];               /* escape sequence string */
int linenum;                   /* current line number in input file */
int tunenum;                   /* number of current tune */
int tnum1,tnum2;
int numtitle;                  /* how many titles were read */
int mline;                     /* number music lines in current tune */
int nsym;                      /* number of symbols in line */
int nsym0;                     /* nsym at start of parsing a line */
int pagenum;                   /* current page in output file */
int writenum;                  /* calls to write_buffer for each one tune */
int xrefnum;                   /* xref number of current tune */
int do_meter, do_indent;       /* how to start next block */

int index_pagenum;             /* for index file */
float index_posx, index_posy;
int index_initialized;

GchordList prep_gchlst;          /* guitar chords for preparsing */
Deco prep_deco;                  /* decorations for preparsing */
int bagpipe;                     /* switch for HP mode */
int within_tune, within_block;   /* where we are in the file */
int do_this_tune;                /* are we typesetting the current one ? */
float posx,posy;                 /* overall scale, position on page */
int barinit;                     /* carryover bar number between parts */

char *p, *p0;                    /* global pointers for parsing music line */

int word,slur;                   /* variables used for parsing... */
int last_note,last_real_note;
int pplet,qplet,rplet;
int carryover;                   /* for interpreting > and < chars */
int ntinext,tinext[MAXHD];       /* for chord ties */

struct ENDINGS ending[20];       /* where to draw endings */           
int num_ending;                  /* number of endings to draw */
int mes1,mes2;                   /* to count measures in an ending */

int slur1[20],slur2[20];         /* needed for drawing slurs */
int overfull;                    /* flag if staff overfull */
int do_words;                    /* flag if staff has words under it */

int vb, verbose;                 /* verbosity, global and within tune */
int in_page=0;

                                 /* switches modified by flags: */
int gmode;                         /* switch for glue treatment */
int include_xrefs;                 /* to include xref numbers in title */
int one_per_page;                  /* new page for each tune ? */
int pagenumbers;                   /* write page numbers ? */
int write_history;                 /* write history and notes ? */
int help_me;                       /* need help ? */
int select_all;                    /* select all tunes ? */
int epsf;                          /* for EPSF postscript output */
int choose_outname;                /* 1 names outfile w. title/fnam */
int break_continues;               /* ignore continuations ? */
int search_field0;                 /* default search field */
int pretty;                        /* for pretty but sprawling layout */
int bars_per_line;                 /* bars for auto linebreaking */
int continue_lines;                /* flag to continue all lines */
int landscape;                     /* flag for landscape output */
int barnums;                       /* interval for bar numbers */
bool make_index;                   // write index file
int notab;                         /* do not process tablature */
int transposegchords;              /* transpose gchords */
float alfa_c;                      /* max compression allowed */
float scalefac;                    /* scale factor for symbol size */
float lmargin;                     /* left margin */
float swidth;                      /* staff width */
float staffsep,dstaffsep;        /* staff separation */
float strict1,strict2;           /* 1stave, mstave strictness */
char transpose[21];              /* target key for transposition */
struct PAPERSIZE* paper;         /* paper size (a4, letter) */
int noauthor;                    /* suppress PS author tag (%%For) */


float alfa_last,beta_last;       /* for last short short line.. */

char in_file[MAXINF][STRLFILE];  /* list of input file names */
signed int  number_input_files;  /* number of input file names */
FILE *fin;                       /* for input file */

char outf[STRLFILE];             /* output file name */
char outfnam[STRLFILE];          /* internal file name for open/close */
char styf[STRLFILE];             /* layout style file name */
char styd[STRLFILE];             /* layout style directory */
char infostr[STRLFILE];          /* title string in PS file */

int  file_open;                  /* for output file */
int  file_initialized;           /* for output file */
FILE *fout,*findex;              /* for output file */
int nepsf;                       /* counter for epsf output files */

char sel_str[MAXINF][STRLFILE];  /* list of selector strings */
int  s_field[MAXINF];            /* type of selection for each file */
int  psel[MAXINF];               /* pointers from files to selectors */

int temp_switch;

/* signal handler for premature termination */
void sighnd_term(int sig)
{
  close_output_file();
  exit(130);
}


/* ----- start of main ------ */
int main(int argc, char **argv)
{
  char ext[41];
  char xref_str[STRLFILE], pat[30][STRLFILE];
  int isel,j,npat,search_field,retcode,rc;

  /* cleanup on premature termination */
  if ((signal(SIGINT,sighnd_term)==SIG_ERR) ||
      (signal(SIGTERM,sighnd_term)==SIG_ERR)) {
    warning("could not install signal handler for SIGTERM and SIGINT","");
  }

  /* ----- set default options and parse arguments ----- */

  maxSyms = allocSyms = 800;
  maxVc   = allocVc = 3;

  init_ops(true);   // do_output = false
  retcode=parse_args (argc, argv);   // do_output = true
  if (retcode) { exit (1); }
  if (do_output) {
    printf("This is abctab2ps, version %s.%s (%s)\n", VERSION, REVISION, VDATE);
  }

  alloc_structs ();

  /* ----- set the page format ----- */
  nfontnames=0;
  if (!set_page_format()) exit (3);
  if (help_me==2) {
    print_format(cfmt);
    exit (0);
  }
  
  /* ----- help printout  ----- */
  if (help_me==1) {
    write_help ();
    exit (0);
  }
 
  if (number_input_files > 0) {
      isel = psel[number_input_files-1];
  } else { 
      isel=psel[0]; 
  } 
  search_field0=s_field[isel];   /* default for interactive mode */
  if (epsf) cutext(outf);
  
  /* ----- initialize ----- */
  zero_sym();
  pagenum=0;
  tunenum=tnum1=tnum2=0;
  verbose=0;
  file_open=file_initialized=0;
  nepsf=0;
  bposy=0;
  posx=cfmt.leftmargin;
  posy=cfmt.pageheight-cfmt.topmargin; 

  strcpy(page_init, "");

  std::cout << "do_output: " << do_output << "\n";
  std::cout << "make_index: " << make_index << "\n";
  if (do_output && make_index) { open_index_file (INDEXFILE); } 

  /* ----- loop over files in list ---- */
  if (number_input_files==0) printf ("++++ No input files, read from stdin\n");
  for (j=0;j<=number_input_files;j++) {
    if (number_input_files==0) {
      /* no input file specified: open stdin */
      fin = stdin;
      strcpy(in_file[0],"stdin");
    } else {
      /* process list of input files */
      if (j==number_input_files) {
        break;
      } else {
        getext (in_file[j],ext);
        /*  skip .ps and .eps files */
        if ((!strcmp(ext,"ps"))||(!strcmp(ext,"eps"))) continue;

        if ((fin = fopen (in_file[j],"r")) == NULL) {      
          if (!strcmp(ext,"")) strext (in_file[j],in_file[j],"abc",1);
          if ((fin = fopen (in_file[j],"r")) == NULL) { 
            printf ("++++ Cannot open input file: %s\n", in_file[j]);
            continue;
          }
        }
      }
    }

    isel=psel[j];
    search_field=s_field[isel];
    npat=rehash_selectors (sel_str[isel], xref_str, pat);
    dfmt=sfmt;
    strcpy(infostr, in_file[j]);

    // The code in broken here as do_output is forever true.
    if (not do_output) {
      printf ("%s:\n", in_file[j]);
      do_index (fin,xref_str,npat,pat,select_all,search_field);
    } else {
      if (!epsf) {
        strext (outf, outf, "ps", 1);
        if (choose_outname) strext (outf, in_file[j], "ps", 1);
        open_output_file(outf,in_file[j]);
      }
      printf ("%s: ", in_file[j]); 
      if (vb>=3) printf ("\n"); 
      process_file (fin,fout,xref_str,npat,pat,select_all,search_field);
      printf ("\n");
    }
  }
  
  if (not do_output)
    printf ("Selected %d title%s of %d\n", tnum1, tnum1==1?"":"s", tnum2);
  
  if (do_output && make_index) close_index_file ();
  rc = close_output_file ();

  if (do_output && rc)
    return 1;
  else
    return 0;
}

